synapse/types.py

```diff
@@ -625,6 +625,22 @@ async def to_string(self, store: "DataStore") -> str:
             return "s%d" % (self.stream,)
 
 
+class StreamKeyType:
+    """Known stream types.
+
+    A stream is a list of entities ordered by an incrementing "stream token".
+    """
```

Comment:
a quick driveby: we should probably give more context around where these "streams" are used. Is it somethign to do with replication, or the database or is it entirely internal? If the latter, where is it used? (honestly, I've entirely paged this out. Sorry!)

Reply:
They're entirely internal, but quite central to Synapse - hence the slightly vague description here. One could write a lot about streams! And the documentation may be a better place for it.

Stream contents are stored in the database, and each stream is its own table with a stream token (an incrementing integer), and then other columns related to the stream type. Different internal systems processing the stream can then easily determine to what point in the stream they've processed up to. As an example, the sync token contains a list of stream tokens separated by underscores. This helps indicate to Synapse at which point in each stream that client has information about.

Another example is when sending events to application services. A token per stream type per appservice is recorded in order to determine what entities (PDUs, EDUs) still need to be pushed to that appservice (any stream entries with a token greater than the saved token).

I would like to document streams in greater detail... but I might do that in a separate area and keep this docstring on just the types light.

Reply:
well, I won't block this PR for it, but I would really love to see a few more words here.